base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,443 @@
+ *              Copyright 2023 Ethan Rodensky <splitwirez@gmail.com>
+ */
+
+#include "traydeskbtn.h"
+#include <limits.h>
+
+#define IDI_SHELL32_DESKTOP 35
+#define IDI_IMAGERES_DESKTOP 110
+
+#define SHOW_DESKTOP_TIMER_ID 999
+#define SHOW_DESKTOP_TIMER_INTERVAL 200
+
+CTrayShowDesktopButton::CTrayShowDesktopButton() :
+    m_nClickedTime(0),
+    m_inset({2, 2}),
+    m_icon(NULL),
+    m_highContrastMode(FALSE),
+    m_drawWithDedicatedBackground(FALSE),
+    m_bHovering(FALSE),
+    m_bPressed(FALSE),
+    IsHorizontal(FALSE)
+{
+}
+
+BOOL
+CTrayShowDesktopButton::GetTaskbar(OUT HWND* taskbarWnd)
+{
+    *taskbarWnd = NULL;
+    auto parent = GetParent();
+    if (!parent)
+        return FALSE;
+
+    parent = parent.GetParent();
+    if (!parent)
+        return FALSE;
+
+    HWND wnd = parent.m_hWnd;
+    *taskbarWnd = wnd;
+    return wnd != NULL;
+}
+
+INT CTrayShowDesktopButton::WidthOrHeight() const
+{
+    if (IsThemeActive() && !m_highContrastMode)
+    {
+        if (m_drawWithDedicatedBackground)
+        {
+            if (GetSystemMetrics(SM_TABLETPC))
+            {
+                //TODO: DPI scaling - return logical-to-physical conversion of 24, not fixed value
+                return 24;
+            }
+            else
+                return 15;
+        }
+        else
+        {
+            return max(16 + (IsHorizontal
+                ? (m_ContentMargins.cxLeftWidth + m_ContentMargins.cxRightWidth)
+                : (m_ContentMargins.cyTopHeight + m_ContentMargins.cyBottomHeight)
+            ), 18) + 6;
+        }
+    }
+    else
+    {
+        return max(
+            2 * GetSystemMetrics(SM_CXBORDER) + GetSystemMetrics(SM_CXSMICON),
+            2 * GetSystemMetrics(SM_CYBORDER) + GetSystemMetrics(SM_CYSMICON)
+        );
+    }
+}
+
+HRESULT CTrayShowDesktopButton::DoCreate(HWND hwndParent)
+{
+    DWORD style = WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | BS_DEFPUSHBUTTON;
+    Create(hwndParent, NULL, NULL, style);
+
+    if (!m_hWnd)
+        return E_FAIL;
+
+    bool bIconRetrievalFailed = ExtractIconExW(L"imageres.dll", -IDI_IMAGERES_DESKTOP, NULL, &m_icon, 1) == UINT_MAX;
+    if (bIconRetrievalFailed || !m_icon)
+        ExtractIconExW(L"shell32.dll", -IDI_SHELL32_DESKTOP, NULL, &m_icon, 1);
+    
+    m_szIcon.cx = 16;
+    m_szIcon.cy = 16;
+    
+    if (m_icon)
+    {
+        ICONINFO iconInfo;
+        if (GetIconInfo(m_icon, &iconInfo))
+        {
+            BITMAP bmpIcon;
+            if (GetObjectW(iconInfo.hbmMask, sizeof(bmpIcon), &bmpIcon))
+            {
+                m_szIcon.cx = bmpIcon.bmWidth;
+                m_szIcon.cy = bmpIcon.bmHeight;
+                /*
+                If this structure defines a black and white icon, this bitmask is formatted
+                so that the upper half is the icon AND bitmask and the lower half is
+                the icon XOR bitmask.
+                */
```

Comment:
```suggestion
                 * If this structure defines a black and white icon, this bitmask is formatted
                 * so that the upper half is the icon AND bitmask and the lower half is
                 * the icon XOR bitmask.
                 */
```
or maybe just use the `//` style of C++ comment.